File: ts/src/registry.ts
import { FacetError, InvariantError, UnknownKindError, ValidationError } from "./errors.js";
import type { Node } from "./node.js";

export type Invariant = (node: Node) => void;

export interface ShapeSpec {
  name: string;
  requiredFacets?: Set<string>;
  optionalFacets?: Set<string>;
  invariants?: Invariant[];
}

export interface KindSpec {
  name: string;
  shape?: string;
  requiredFacets?: Set<string>;
  optionalFacets?: Set<string>;
  invariants?: Invariant[];
}

/** One structured validation finding from `Registry.check` (never thrown). */
export interface Violation {
  readonly code: string;
  readonly detail: string;
  readonly message: string;
}

export class Registry {
  private specs = new Map<string, KindSpec>();
  private shapes = new Map<string, ShapeSpec>();

  registerShape(spec: ShapeSpec): void {
    if (this.shapes.has(spec.name)) {
      throw new ValidationError(`shape ${JSON.stringify(spec.name)} is already registered`);
    }
    this.shapes.set(spec.name, spec);
  }

  isShape(name: string): boolean {
    return this.shapes.has(name);
  }

  register(spec: KindSpec): void {
    if (this.specs.has(spec.name)) {
      throw new ValidationError(`kind ${JSON.stringify(spec.name)} is already registered`);
    }
    if (spec.shape !== undefined && !this.shapes.has(spec.shape)) {
      throw new UnknownKindError(
        `kind ${JSON.stringify(spec.name)} adopts unknown shape ${JSON.stringify(spec.shape)}`,
      );
    }
    this.specs.set(spec.name, spec);
  }

  isRegistered(kind: string): boolean {
    return this.specs.has(kind);
  }

  get(kind: string): KindSpec {
    const spec = this.specs.get(kind);
    if (spec === undefined) {
      throw new UnknownKindError(`kind ${JSON.stringify(kind)} is not registered`);
    }
    return spec;
  }

  private compose(
    spec: KindSpec,
    kind: string,
  ): { required: Set<string>; optional: Set<string>; invariants: Invariant[] } {
    const required = new Set(spec.requiredFacets ?? []);
    const optional = new Set(spec.optionalFacets ?? []);
    const invariants: Invariant[] = [];
    if (spec.shape !== undefined) {
      const shape = this.shapes.get(spec.shape);
      if (shape === undefined) {
        throw new UnknownKindError(`kind ${JSON.stringify(kind)} adopts unknown shape ${JSON.stringify(spec.shape)}`);
      }
      for (const f of shape.requiredFacets ?? []) required.add(f);
      for (const f of shape.optionalFacets ?? []) optional.add(f);
      for (const inv of shape.invariants ?? []) invariants.push(inv);
    }
    for (const inv of spec.invariants ?? []) invariants.push(inv);
    return { required, optional, invariants };
  }

  validate(node: Node): void {
    const spec = this.get(node.kind);
    const { required, optional, invariants } = this.compose(spec, node.kind);
    const present = new Set(Object.keys(node.facets));
    const missing = [...required].filter((f) => !present.has(f)).sort();
    if (missing.length > 0) {
      throw new FacetError(`${node.id}: missing required facets ${JSON.stringify(missing)}`);
    }
    const allowed = new Set([...required, ...optional]);
    const unexpected = [...present].filter((f) => !allowed.has(f)).sort();
    if (unexpected.length > 0) {
      throw new FacetError(`${node.id}: unexpected facets ${JSON.stringify(unexpected)}`);
    }
    for (const invariant of invariants) invariant(node);
  }

  /** Collect ALL violations of `node` with machine-stable codes; never throws on
   * content. Non-kernel exceptions from invariants are programmer bugs and propagate. */
  check(node: Node): Violation[] {
    const spec = this.specs.get(node.kind);
    if (spec === undefined) {
      return [
        {
          code: "unknown-kind",
          detail: node.kind,
          message: `${node.id}: kind ${JSON.stringify(node.kind)} is not registered`,
        },
      ];
    }
    const { required, optional, invariants } = this.compose(spec, node.kind);
    const present = new Set(Object.keys(node.facets));
    const violations: Violation[] = [];
    for (const name of [...required].filter((f) => !present.has(f)).sort()) {
      violations.push({
        code: "facet-missing",
        detail: name,
        message: `${node.id}: missing required facet ${JSON.stringify(name)}`,
      });
    }
    const allowed = new Set([...required, ...optional]);
    for (const name of [...present].filter((f) => !allowed.has(f)).sort()) {
      violations.push({
        code: "facet-unexpected",
        detail: name,
        message: `${node.id}: unexpected facet ${JSON.stringify(name)}`,
      });
    }
    // Invariants presuppose their facets; running them anyway would duplicate reports.
    if (violations.length > 0) return violations;
    for (const invariant of invariants) {
      try {
        invariant(node);
      } catch (exc) {
        if (exc instanceof FacetError) {
          violations.push({ code: "facet-invalid", detail: "", message: exc.message });
        } else if (exc instanceof InvariantError) {
          violations.push({ code: "invariant-violated", detail: "", message: exc.message });
        } else {
          throw exc;
        }
      }
    }
    return violations;
  }
}

File: python/src/nodes/core/registry.py
from __future__ import annotations

from typing import Callable

from pydantic import BaseModel, ConfigDict, Field

from nodes.core.errors import FacetError, InvariantError, UnknownKindError, ValidationError
from nodes.core.node import Node

Invariant = Callable[[Node], None]


class ShapeSpec(BaseModel):
    model_config = ConfigDict(arbitrary_types_allowed=True)

    name: str
    required_facets: set[str] = Field(default_factory=set)
    optional_facets: set[str] = Field(default_factory=set)
    invariants: list[Invariant] = Field(default_factory=list)


class KindSpec(BaseModel):
    model_config = ConfigDict(arbitrary_types_allowed=True)

    name: str
    shape: str | None = None
    required_facets: set[str] = Field(default_factory=set)
    optional_facets: set[str] = Field(default_factory=set)
    invariants: list[Invariant] = Field(default_factory=list)


class Violation(BaseModel):
    """One structured validation finding from `Registry.check` (never raised)."""

    code: str
    detail: str
    message: str


class Registry:
    def __init__(self) -> None:
        self._specs: dict[str, KindSpec] = {}
        self._shapes: dict[str, ShapeSpec] = {}

    def register_shape(self, spec: ShapeSpec) -> None:
        if spec.name in self._shapes:
            raise ValidationError(f"shape {spec.name!r} is already registered")
        self._shapes[spec.name] = spec

    def is_shape(self, name: str) -> bool:
        return name in self._shapes

    def register(self, spec: KindSpec) -> None:
        if spec.name in self._specs:
            raise ValidationError(f"kind {spec.name!r} is already registered")
        if spec.shape is not None and spec.shape not in self._shapes:
            raise UnknownKindError(f"kind {spec.name!r} adopts unknown shape {spec.shape!r}")
        self._specs[spec.name] = spec

    def is_registered(self, kind: str) -> bool:
        return kind in self._specs

    def get(self, kind: str) -> KindSpec:
        try:
            return self._specs[kind]
        except KeyError as exc:
            raise UnknownKindError(f"kind {kind!r} is not registered") from exc

    def _compose(self, spec: KindSpec) -> tuple[set[str], set[str], list[Invariant]]:
        required = set(spec.required_facets)
        optional = set(spec.optional_facets)
        invariants: list[Invariant] = []
        if spec.shape is not None:
            shape = self._shapes[spec.shape]
            required |= shape.required_facets
            optional |= shape.optional_facets
            invariants.extend(shape.invariants)
        invariants.extend(spec.invariants)
        return required, optional, invariants

    def validate(self, node: Node) -> None:
        required, optional, invariants = self._compose(self.get(node.kind))
        present = set(node.facets)
        missing = required - present
        if missing:
            raise FacetError(f"{node.id}: missing required facets {sorted(missing)}")
        allowed = required | optional
        unexpected = present - allowed
        if unexpected:
            raise FacetError(f"{node.id}: unexpected facets {sorted(unexpected)}")
        for invariant in invariants:
            invariant(node)

    def check(self, node: Node) -> list[Violation]:
        """Collect ALL violations of `node` with machine-stable codes; never raises on
        content. Non-kernel exceptions from invariants are programmer bugs and propagate."""
        spec = self._specs.get(node.kind)
        if spec is None:
            return [
                Violation(
                    code="unknown-kind",
                    detail=node.kind,
                    message=f"{node.id}: kind {node.kind!r} is not registered",
                )
            ]
        required, optional, invariants = self._compose(spec)
        present = set(node.facets)
        violations: list[Violation] = []
        for name in sorted(required - present):
            violations.append(
                Violation(code="facet-missing", detail=name, message=f"{node.id}: missing required facet {name!r}")
            )
        for name in sorted(present - (required | optional)):
            violations.append(
                Violation(code="facet-unexpected", detail=name, message=f"{node.id}: unexpected facet {name!r}")
            )
        if violations:
            return violations  # invariants presuppose their facets; running them would duplicate reports
        for invariant in invariants:
            try:
                invariant(node)
            except FacetError as exc:
                violations.append(Violation(code="facet-invalid", detail="", message=str(exc)))
            except InvariantError as exc:
                violations.append(Violation(code="invariant-violated", detail="", message=str(exc)))
        return violations
